xfer/app_client.go

```diff
@@ -0,0 +1,166 @@
+	ControlConnection(handler ControlHandlerFunc)
+	Stop()
+}
+
+type appClient struct {
+	ProbeConfig
+
+	quit     chan struct{}
+	target   string
+	insecure bool
+	client   http.Client
+
+	controlServerCodecMtx sync.Mutex
+	controlServerCodec    rpc.ServerCodec
+}
+
+// NewAppClient makes a new AppClient.
+func NewAppClient(pc ProbeConfig, hostname, target string) (AppClient, error) {
+	httpTransport, err := getHTTPTransport(hostname, pc.Insecure)
+	if err != nil {
+		return nil, err
+	}
+
+	return &appClient{
+		ProbeConfig: pc,
+		quit:        make(chan struct{}),
+		target:      target,
+		client: http.Client{
+			Transport: httpTransport,
+		},
+	}, nil
+}
+
+func (c *appClient) Stop() {
+	c.controlServerCodecMtx.Lock()
+	defer c.controlServerCodecMtx.Unlock()
+	close(c.quit)
+	if c.controlServerCodec != nil {
+		c.controlServerCodec.Close()
+	}
+}
+
+// AuthorizationHeader returns a value suitable for an HTTP Authorization
+// header, based on the passed token string.
+func AuthorizationHeader(token string) string {
+	return fmt.Sprintf("Scope-Probe token=%s", token)
+}
+
+func (c *appClient) authoriseHeaders(headers http.Header) {
+	headers.Set("Authorization", AuthorizationHeader(c.Token))
+	headers.Set(ScopeProbeIDHeader, c.ProbeID)
+}
+
+func (c *appClient) authorizedRequest(method string, urlStr string, body io.Reader) (*http.Request, error) {
+	req, err := http.NewRequest(method, urlStr, body)
+	if err == nil {
+		c.authoriseHeaders(req.Header)
+	}
+	return req, err
+}
+
+func (c *appClient) Details() (Details, error) {
+	result := Details{}
+	req, err := c.authorizedRequest("GET", sanitize.URL("", 0, "/api")(c.target), nil)
+	if err != nil {
+		return result, err
+	}
+	resp, err := c.client.Do(req)
+	if err != nil {
+		return result, err
+	}
+	defer resp.Body.Close()
+	return result, json.NewDecoder(resp.Body).Decode(&result)
+}
+
+func (c *appClient) controlConnection(handler ControlHandlerFunc) error {
+	dialer := websocket.Dialer{}
+	headers := http.Header{}
+	c.authoriseHeaders(headers)
+	// TODO(twilkie) need to update sanitize to work with wss
+	url := sanitize.URL("ws://", 0, "/api/controls")(c.target)
+	conn, _, err := dialer.Dial(url, headers)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		log.Printf("Closing control connection to %s", c.target)
+		conn.Close()
+	}()
+
+	codec := NewJSONWebsocketCodec(conn)
+	server := rpc.NewServer()
+	if err := server.RegisterName("control", handler); err != nil {
+		return err
+	}
+
+	c.controlServerCodecMtx.Lock()
+	c.controlServerCodec = codec
+	// At this point we may have tried to quit earlier, so check to see if the
+	// quit channel has been closed, non-blocking.
+	select {
+	default:
+	case <-c.quit:
+		codec.Close()
+		return nil
+	}
+	c.controlServerCodecMtx.Unlock()
+
+	server.ServeCodec(codec)
+	return nil
+}
+
+func (c *appClient) ControlConnection(handler ControlHandlerFunc) {
+	backoff := initialBackoff
+
+	for {
+		err := c.controlConnection(handler)
```

Comment:
Maybe a better name for this method, that describes what it does or what it returns? `setupConnection`? And, if it were to return the servercodec, would we still need the mutex? or could we contain the codec to this method, and clean it up when quit is closed?